File: java/iot3/mobility/src/main/java/com/orange/iot3mobility/roadobjects/RoadUser.java
/*
 Copyright 2016-2024 Orange

 This software is distributed under the MIT license, see LICENSE.txt file for more details.

 @author Mathieu LEFEBVRE <mathieu1.lefebvre@orange.com>
 */
package com.orange.iot3mobility.roadobjects;

import com.orange.iot3mobility.messages.StationType;
import com.orange.iot3mobility.messages.cam.core.CamCodec;
import com.orange.iot3mobility.quadkey.LatLng;

public class RoadUser {

    private static final int LIFETIME = 1500; // 1.5 seconds

    private final String uuid;
    private StationType stationType;
    private LatLng position;
    private double speed; // m/s
    private double heading; // degree
    private long timestamp;
    private Long linkedStationId;
    private CamCodec.CamFrame<?> camFrame;

    public RoadUser(String uuid, StationType stationType, LatLng position, double speed, double heading,
                    CamCodec.CamFrame<?> camFrame) {
        this(uuid, stationType, position, speed, heading, null, camFrame);
    }

    public RoadUser(String uuid, StationType stationType, LatLng position, double speed, double heading,
                    Long linkedStationId, CamCodec.CamFrame<?> camFrame) {
        this.uuid = uuid;
        this.setStationType(stationType);
        this.position = position;
        this.speed = speed;
        this.heading = heading;
        this.linkedStationId = linkedStationId;
        this.camFrame = camFrame;
        updateTimestamp();
    }

    public String getUuid() {
        return uuid;
    }

    public StationType getStationType() {
        return stationType;
    }

    public void setStationType(StationType stationType) {
        this.stationType = stationType;
    }

    public LatLng getPosition() {
        return position;
    }

    public void setPosition(LatLng position) {
        this.position = position;
    }

    public double getSpeed() {
        return speed;
    }

    public double getSpeedKmh() {
        return speed * 3.6;
    }

    public void setSpeed(double speed) {
        this.speed = speed;
    }

    public double getHeading() {
        return heading;
    }

    public void setHeading(double heading) {
        this.heading = heading;
    }

    public void setCamFrame(CamCodec.CamFrame<?> camFrame) {
        this.camFrame = camFrame;
    }

    public CamCodec.CamFrame<?> getCamFrame() {
        return camFrame;
    }

    /**
     * Returns the linked station identifier, or {@code null} if not linked.
     */
    public Long getLinkedStationId() {
        return linkedStationId;
    }

    public void setLinkedStationId(Long linkedStationId) {
        this.linkedStationId = linkedStationId;
    }

    /**
     * Returns {@code true} if this road user has a linked station identifier
     * (e.g. a trailer or a platooning pair).
     *
     * @see com.orange.iot3mobility.managers.RoadUserManager#getLinkedObject(RoadUser)
     */
    public boolean isLinked() {
        return linkedStationId != null;
    }

    public long getTimestamp() {
        return timestamp;
    }

    public void updateTimestamp() {
        this.timestamp = System.currentTimeMillis();
    }

    public boolean stillLiving() {
        return System.currentTimeMillis() - timestamp < LIFETIME;
    }

    public boolean isVulnerable() {
        return stationType.equals(StationType.CYCLIST) || stationType.equals(StationType.PEDESTRIAN);
    }

}

File: swift/ITSClient/Sources/ITSMobility/Observers/RoadUser/RoadUser.swift
/*
 * Software Name : ITSClient
 * SPDX-FileCopyrightText: Copyright (c) Orange SA
 * SPDX-License-Identifier: MIT
 *
 * This software is distributed under the MIT license,
 * see the "LICENSE.txt" file for more details or https://opensource.org/license/MIT/
 *
 * Software description: Swift ITS client.
 */

import Foundation

/// The representation of a road user.
public struct RoadUser: Sendable {
    /// The unique identifier.
    public let id: String
    /// The station type.
    public let stationType: StationType?
    /// The longitude.
    public let longitude: Double
    /// The latitude.
    public let latitude: Double
    /// The speed in m/s.
    public let speed: Double?
    /// The heading in degrees.
    public let heading: Double?
    /// The road user timestamp.
    public let timestamp: Date
    /// The underlying CAM object.
    public let underlyingCAM: CAM

    var expirationDate: Date {
        // Expiration 1.5s after
        timestamp.addingTimeInterval(1.5)
    }
    
    /// Initializes a `RoadUser`.
    /// - Parameters:
    ///   - id: The unique identifier.
    ///   - stationType: The station type.
    ///   - longitude: The longitude.
    ///   - latitude: The latitude.
    ///   - speed: The speed in m/s.
    ///   - heading: The heading in degrees.
    ///   - timestamp: The road user timestamp.
    ///   - underlyingCAM: The underlying CAM object.
    public init(
        id: String,
        stationType: StationType?,
        longitude: Double,
        latitude: Double,
        speed: Double?,
        heading: Double?,
        timestamp: Date,
        underlyingCAM: CAM
    ) {
        self.id = id
        self.stationType = stationType
        self.longitude = longitude
        self.latitude = latitude
        self.speed = speed
        self.heading = heading
        self.timestamp = timestamp
        self.underlyingCAM = underlyingCAM
    }
}
